// jwks-rsa/ts-output/ts-definitions.tests.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var nock_1 = __importDefault(require("nock"));
var jwksRsa = require('../src');
describe('typescript definition', function () {
    var jwksHost = 'http://my-authz-server';
    var givenPublicCertOnAuthzServer = function (kid, cert) {
        nock_1.default(jwksHost)
            .get('/.well-known/jwks.json')
            .reply(200, {
            keys: [
                {
                    alg: 'RS256',
                    kty: 'RSA',
                    use: 'sig',
                    x5c: [cert],
                    kid: kid
                }
            ]
        });
    };
    describe('hapiJwt2KeyAsync', function () {
        // it('should return a secret provider function', async () => {
        //   givenPublicCertOnAuthzServer('someKeyId', 'pk1');
        //   const secretProvider = jwksRsa.hapiJwt2KeyAsync({
        //     jwksUri: `${jwksHost}/.well-known/jwks.json`
        //   });
        //   const { key } = await secretProvider({
        //     header: {
        //       'alg': 'RS256',
        //       'kid': 'someKeyId'
        //     }
        //   });
        //   expect(key).to.contain('pk1');
        // });
    });
});
